src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityAnalyzer.ImmutableTypeParameterArgumentAnalysis.cs

```diff
@@ -0,0 +1,642 @@
+						ctx.ReportDiagnostic,
+						annotationsContext,
+						immutabilityContext,
+						operation.NarrowedType,
+						() => ( (TypePatternSyntax)operation.Syntax ).Type
+					);
+				},
+				OperationKind.TypePattern
+			);
+
+			// type arguments of types in:
+			//   "foo is not T x"
+			//   cast T x:        (switch statement)
+			//   T x =>           (switch expression)
+			context.RegisterOperationAction(
+				ctx => {
+					var operation = (IDeclarationPatternOperation)ctx.Operation;
+
+					AnalyzeTypeRecursive(
+						ctx.ReportDiagnostic,
+						annotationsContext,
+						immutabilityContext,
+						operation.NarrowedType,
+						() => ( (DeclarationPatternSyntax)operation.Syntax ).Type
+					);
+				},
+				OperationKind.DeclarationPattern
+			);
+
+			// type arguments of types in casts:
+			//   (T)foo
+			//   foo as T
+			context.RegisterOperationAction(
+				ctx => {
+					var operation = (IConversionOperation)ctx.Operation;
+
+					if( operation.IsImplicit ) {
+						return;
+					}
+
+					SyntaxNodeOrToken getSyntax() => operation.Syntax switch {
+						BinaryExpressionSyntax binaryExpression => binaryExpression.Right,
+						CastExpressionSyntax castExpression => castExpression.Type,
+						_ => operation.Syntax
+					};
+
+					AnalyzeTypeRecursive(
+						ctx.ReportDiagnostic,
+						annotationsContext,
+						immutabilityContext,
+						operation.Type!,
+						getSyntax
+					);
+				},
+				OperationKind.Conversion
+			);
+		}
+
+		private static void AnalyzeMethodUsage(
+			OperationAnalysisContext ctx,
+			AnnotationsContext annotationsContext,
+			ImmutabilityContext immutabilityContext,
+			IMethodSymbol method,
+			Func<SyntaxNodeOrToken> getAnalyzedSyntax
+		) {
+			AnalyzeTypeArgumentsRecursive(
+				ctx.ReportDiagnostic,
+				annotationsContext,
+				immutabilityContext,
+				method.TypeArguments,
+				method.TypeParameters,
+				getAnalyzedSyntax
+			);
+
+			AnalyzeMaybeStaticMemberAccess(
+				ctx,
+				annotationsContext,
+				immutabilityContext,
+				method,
+				getAnalyzedSyntax
+			);
+		}
+
+		private static void AnalyzeMaybeStaticMemberAccess(
+			OperationAnalysisContext ctx,
+			AnnotationsContext annotationsContext,
+			ImmutabilityContext immutabilityContext,
+			ISymbol member,
+			Func<SyntaxNodeOrToken> getAnalyzedSyntax
+		) {
+			if( !member.IsStatic ) {
+				return;
+			}
+
+			AnalyzeTypeRecursive(
+				ctx.ReportDiagnostic,
+				annotationsContext,
+				immutabilityContext,
+				member.ContainingType,
+				SelectLeftSyntax( getAnalyzedSyntax )
+			);
+		}
+
+		private static void AnalyzeTypeRecursive(
+			Action<Diagnostic> reportDiagnostic,
+			AnnotationsContext annotationsContext,
+			ImmutabilityContext immutabilityContext,
+			ITypeSymbol type,
+			Func<SyntaxNodeOrToken> getAnalyzedSyntax
+		) {
+			INamedTypeSymbol? namedType = GetNamedTypeRecursive( type );
+			if( namedType is null ) {
+				return;
+			}
+
+			if( namedType.ContainingType is not null ) {

```

Comment:
Considered adding an `if( !namedType.IsGenericType ) { return; }` here, but that iterates up the containing types checking for `TypeParameters.Length != 0` anyway, so, it wouldn't be that much cheaper than what's ultimately happening, it would save a couple closures, but also add extra cost for actual generic things.

https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/NamedTypeSymbol.cs,89e394ddef6358d1,references